File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.js
function solution(numbers, hand) {
    
    //키패드
    const keypad = {
        1: [0, 0], 2: [0, 1], 3: [0, 2],
        4: [1, 0], 5: [1, 1], 6: [1, 2],
        7: [2, 0], 8: [2, 1], 9: [2, 2],
        '*': [3, 0], 0: [3, 1], '#': [3, 2]
    };
    
    let answer = [];
    
    let leftHand = keypad['*'];
    let rightHand = keypad['#'];
    
    for(let num of numbers) {
        if([1,4,7].includes(num)) {
            answer.push('L');
            leftHand = keypad[num];
        }
        else if([3,6,9].includes(num)) {
            answer.push('R');
            rightHand = keypad[num];
        } else {
            let leftDist = Math.abs(leftHand[0] - keypad[num][0]) + Math.abs(leftHand[1] - keypad[num][1]);
            let rightDist = Math.abs(rightHand[0] - keypad[num][0]) + Math.abs(rightHand[1] - keypad[num][1]);
            
            if(leftDist < rightDist) {
                answer.push('L');
                leftHand=keypad[num];
            } else if (leftDist > rightDist) {
                answer.push('R');
                rightHand=keypad[num];
            } else {
                //같은 경우
                if(hand === 'left') {
                    answer.push('L');
                leftHand=keypad[num];
                } else {
                      answer.push('R');
                      rightHand=keypad[num];
                }
            }
        }
    } //for문 끝
    
    
    
    return answer.join('');
}

File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.java
import java.util.*;
import java.io.*;

class Solution {
    public String solution(int[] numbers, String hand) {
        
        Map<Integer, int[]> map = new HashMap<>();
        
        map.put(0, new int[]{3,1});
        map.put(1, new int[]{0,0});
        map.put(2, new int[]{0,1});
        map.put(3, new int[]{0,2});
        map.put(4, new int[]{1,0});
        map.put(5, new int[]{1,1});
        map.put(6, new int[]{1,2});
        map.put(7, new int[]{2,0});
        map.put(8, new int[]{2,1});
        map.put(9, new int[]{2,2});
        
        StringBuilder answer = new StringBuilder();
        
        int[] left = {3, 0};
        int[] right = {3, 2};
        
        for(int n = 0; n < numbers.length; n++) {
            
            int[] now = map.get(numbers[n]);
            
            if(numbers[n] % 3 == 1) {
                answer.append("L");
                left[0] = now[0];
                left[1] = now[1];
            } else if(numbers[n] % 3 == 2 || numbers[n] == 0) {
                int diffR = Math.abs(right[0] - now[0]) + Math.abs(right[1] - now[1]);
                int diffL = Math.abs(left[0] - now[0]) + Math.abs(left[1] - now[1]);
                
                if(diffR > diffL) {
                    answer.append("L");
                    left[0] = now[0];
                    left[1] = now[1];
                } else if (diffR < diffL) {
                    answer.append("R");
                    right[0] = now[0];
                    right[1] = now[1];
                } else {
                    if(hand.equals("right")) {
                        answer.append("R");
                        right[0] = now[0];
                        right[1] = now[1];
                    } else {
                        answer.append("L");
                        left[0] = now[0];
                        left[1] = now[1];
                    }
                }
            } else {
                answer.append("R");
                right[0] = now[0];
                right[1] = now[1];
            }
        }
        
        return answer.toString();
    }
}